kernel/tests/read.rs

```diff
@@ -353,35 +356,159 @@ async fn stats() -> Result<(), Box<dyn std::error::Error>> {
+        .into())
+}
+
+fn read_with_execute(
+    engine: &dyn Engine,
+    scan: &Scan,
+    expected: &[String],
+) -> Result<(), Box<dyn std::error::Error>> {
+    let scan_results = scan.execute(engine)?;
+    let batches: Vec<RecordBatch> = scan_results
+        .into_iter()
+        .map(|sr| {
+            let data = sr.raw_data.unwrap();
+            let record_batch = to_arrow(data).unwrap();
+            if let Some(mask) = sr.mask {
+                filter_record_batch(&record_batch, &mask.into()).unwrap()
+            } else {
+                record_batch
+            }
+        })
+        .collect();
+
+    if expected.is_empty() {
+        assert_eq!(batches.len(), 0);
+    } else {
+        let schema = batches[0].schema();
+        let batch = concat_batches(&schema, &batches)?;
+        assert_batches_sorted_eq!(expected, &[batch]);
+    }
+    Ok(())
+}
+
+struct ScanFile {
+    path: String,
+    size: i64,
+    dv_info: DvInfo,
+    partition_values: HashMap<String, String>,
+}
+
+fn scan_data_callback(
+    batches: &mut Vec<ScanFile>,
+    path: &str,
+    size: i64,
+    dv_info: DvInfo,
+    partition_values: HashMap<String, String>,
+) {
+    batches.push(ScanFile {
+        path: path.to_string(),
+        size,
+        dv_info,
+        partition_values,
+    });
+}
+
+fn read_with_scan_data(
+    location: &Url,
+    engine: &dyn Engine,
+    scan: &Scan,
+    expected: &[String],
+) -> Result<(), Box<dyn std::error::Error>> {
+    let global_state = scan.global_scan_state();
+    let scan_data = scan.scan_data(engine)?;
+    let mut scan_files = vec![];
+    for data in scan_data {
+        let (data, vec) = data?;
+        scan_files = visit_scan_files(data.as_ref(), &vec, scan_files, scan_data_callback)?;
+    }
+
+    let mut batches = vec![];
+    for scan_file in scan_files.into_iter() {
+        let file_path = location.join(&scan_file.path)?;
+        let mut selection_vector = scan_file
+            .dv_info
+            .get_selection_vector(engine, location)
+            .unwrap();
+        let meta = FileMeta {
+            last_modified: 0,
+            size: scan_file.size as usize,
+            location: file_path,
+        };
+        let read_results = engine
+            .get_parquet_handler()
+            .read_parquet_files(&[meta], global_state.read_schema.clone(), None)
+            .unwrap();
+
+        for read_result in read_results {
+            let read_result = read_result.unwrap();
+            let len = read_result.length();
+
+            // ask the kernel to transform the physical data into the correct logical form
+            let logical = transform_to_logical(
+                engine,
+                read_result,
+                &global_state,
+                &scan_file.partition_values,
+            )
+            .unwrap();
+
+            let record_batch = to_arrow(logical).unwrap();
+            let rest = selection_vector.as_mut().map(|mask| mask.split_off(len));
```

Comment:
ah selection vector spans the whole parquet file so we need to do chunking?

Reply:
Yep. We don't know how many chunks the reader may have returned, so we need to split up the DV over each chunk (in this case it'll be equivalent, but may as well be future proof)